designs/generics-polymorphism-design.md

```diff
@@ -0,0 +1,210 @@
+The following should print 10, then True:
+```python
+T: TypeVar = TypeVar('T')
+
+class Printer(Generic[T]):
+   def print(self: Printer, x: T):
+       print(x)
+
+p_int: Printer[int] = None
+p_int = Printer[int]()
+p_int.print(10)
+
+p_bool: Printer[bool] = None
+p_bool = Printer[bool]()
+p_bool.print(True)
+```
+
+3. Simple generic class of one object type with invalid print call.
+The following should error, saying something like "trying to call print with invalid arg type (Box)"
+```python
+T: TypeVar = TypeVar('T')
+
+class Box(object):
+    val: int = 10
+
+class Printer(Generic[T]):
+   def print(self: Printer, x: T):
+       print(x)
+
+p: Printer[Box] = None
+p = Printer[Box]()
+p.print(Box())
+```
+
+4. Simple generic class of one object type with invalid binop and generic return.
+The following should error, saying something like "trying to do binop with invalid arg type (bool)"
+```python
+T: TypeVar = TypeVar('T')
+
+class Adder(Generic[T]):
+   def add(self: Printer, x: T, y: T) -> T:
+       return x + y
+
+a: Adder[bool] = None
+a = Adder[bool]()
+print(a.add(True, False))
+```
+
+5. Simple generic class of one object type with valid binop and generic return.
+The following should print 10.
+```python
+T: TypeVar = TypeVar('T')
+
+class Adder(Generic[T]):
+   def add(self: Printer, x: T, y: T) -> T:
+       return x + y
+
+a: Adder[int] = None
+a = Adder[int]()
+print(a.add(4, 6))
+```
+
+6. Simple generic class of one object type with invalid method call.
+The following should error, saying something like "trying to call Printer.print with invalid arg type (Box)"
+```python
+T: TypeVar = TypeVar('T')
+
+class Box(object):
+    val: int = 10
+
+class Printer(Generic[T]):
+   def print(self: Printer, x: T):
+       print(x)
+
+p: Printer[int] = None
+p = Printer[int]()
+p.print(Box())
+```
+
+7. Overlapping generic and identifier names. The following should give an error like "identifier T defined earlier"
+```python
+T: int = 0
+T: TypeVar = TypeVar('T')
+
+class Printer(Generic[T]):
+   def print(self: Printer, x: T):
+       print(x)
+
+p: Printer[int] = None
+p = Printer[int]()
+p.print(10)
+```
+
+8. Overlapping generic and class names. The following should give an error like "unable to resolve between class T and TypeVar T".
+```python
+class T(object):
+    pass
+
+T: TypeVar = TypeVar('T')
+
+class Printer(Generic[T]):
+   def print(self: Printer, x: T):
+       print(x)
+
+p: Printer[int] = None
+p = Printer[int]()
+p.print(10)
+```
+
+9. Generic objects as parameters. The following should print 10.
+```python
+T: TypeVar = TypeVar('T')
+
+class Printer(Generic[T]):
+   def print(self: Printer, x: T):
+       print(x)
+
+def print_ten(p: Printer[int]):
+    p.print(10)
+
+p: Printer[int] = None
+p = Printer[int]()
+print_ten(p)
+```
+
+10. Generic objects as fields. The following should print 10.
+```python
+T: TypeVar = TypeVar('T')
+
+class Printer(Generic[T]):
+   def print(self: Printer, x: T):
+       print(x)
```

Comment:
How about:
`T1: TypeVar = TypeVar('T1')`
`T2: TypeVar = TypeVar('T2')`
`class Printer(Generic[T1,T2])`
`...`
Shouldn't we support this?

Reply:
It should be supported in the code, but I'll add a test for it